### PR TITLE
OpenROAD complete backend with nangate45

### PR DESCRIPTION
**Pull request by ssteffl**
- companion PR to ucb-bar/hammer#584

__overview__
- this assumes a pristine OpenROAD-flow installation somewhere on the machine. no hacks to OpenROAD-flow used 
- the default OpenROAD-flow tcl-scripts are used to drive the tools. hammer just sets up run directories/configs
- documentation not complete, and several TODOs in the code
- ran on personal laptop running Ubuntu 18

__results__
- default RocketConfig makes it through __non-hierarchical__ srams,synthesis,par,drc
- all def/gds outputs look reasonable after each steppnr
- quality issues with drc and PPA reports after routing
- no simulation possible since no stdcell behavioral models exist
- no lvs possible since no stdcell spice/cdl exists
- syn+pnr take ~4 hours, drc might not finish

### CODE REVIEW

File: vlsi/Makefile [hunk 1]
```diff
@@ -19,18 +19,22 @@ include $(base_dir)/variables.mk
...
-endif
+MACROCOMPILER_MODE ?= $(if $(filter $(tech_name),asap7),\
+												--mode synflops,\
+												-l $(SMEMS_CACHE) -hir $(SMEMS_HAMMER) --mode strict)
 ENV_YML            ?= $(vlsi_dir)/env.yml
-INPUT_CONFS        ?= example.yml
-HAMMER_EXEC        ?= ./example-vlsi
+INPUT_CONFS        ?= $(if $(filter $(tech_name),nangate45),\
+												example-nangate45.yml,\
+												example.yml)
```
Comment (colinschmidt) [inline vlsi/Makefile:34]: Perhaps we should name eaxmple.yml example-asap7.yml?

File: vlsi/example-nangate45.yml [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# Technology Setup
+# Technology used is nanagate45
+vlsi.core.technology: nangate45
```
Comment (colinschmidt) [inline vlsi/example-nangate45.yml:3]: We might want to separate out the yaml into generic settings and tech specific settings, so we would have an:
example.yml
example-asap7.yml
example-nangate45.yml

And it optionally includes the right pair based on `tech_name`

File: vlsi/example-vlsi-nangate45 [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+def example_place_tap_cells(x: hammer_vlsi.HammerTool) -> bool:
+    x.append("")
+    return True
+
+def example_add_fillers(x: hammer_vlsi.HammerTool) -> bool:
+    x.append("")
+    return True
+
+class ExampleDriver(CLIDriver):
+    def get_extra_par_hooks(self) -> List[HammerToolHookAction]:
```
Comment (colinschmidt) [inline vlsi/example-vlsi-nangate45:18]: I think we just have a generic exampleDriver that inspects `os.env(tech_name)` and includes the get hooks programmatically.

The magic of hammer 🎉

**Comment by alonamid** [comment]
I know this is still a draft, but I would appreciate seeing docs associated with this.

File: vlsi/Makefile [hunk 4]
```diff
@@ -19,25 +19,27 @@ include $(base_dir)/variables.mk
...
+                        example-nangate45.yml,\
+                        example-asap7.yml)
+HAMMER_EXEC        ?= example-vlsi
 VLSI_TOP           ?= $(TOP)
 VLSI_HARNESS_DUT_NAME ?= dut
 VLSI_OBJ_DIR       ?= $(vlsi_dir)/build
-ifneq ($(CUSTOM_VLOG), )
-	OBJ_DIR        ?= $(VLSI_OBJ_DIR)/custom-$(VLSI_TOP)
+ifneq ($(CUSTOM_VLOG),)
+	OBJ_DIR          ?= $(VLSI_OBJ_DIR)/custom-$(VLSI_TOP)
```
Comment (colinschmidt) [inline vlsi/Makefile:40]: Do these spaces line something up? It seems like they unalign these `?=` operators.

File: vlsi/example-vlsi [hunk 5]
```diff
@@ -7,21 +11,24 @@ from hammer_vlsi import CLIDriver, HammerToolHookAction
 from typing import Dict, Callable, Optional, List
 
 def example_place_tap_cells(x: hammer_vlsi.HammerTool) -> bool:
-    x.append('''
+    if x.get_setting("vlsi.core.technology") == "asap7":
```
Comment (colinschmidt) [inline vlsi/example-vlsi:14]: There is a nicer way to do this now but this is ok for the time being.

**Pull request merged**